### PR TITLE
[SystemZ][XRay] Implement XRay instrumentation for SystemZ

### PR DESCRIPTION
**Pull request by redstar**
Expands pseudo instructions PATCHABLE_FUNCTION_ENTER and PATCHABLE_RET into a small instruction sequence which calls into the XRay library.

### CODE REVIEW

File: llvm/test/CodeGen/SystemZ/xray-01.ll [hunk 1]
```diff
@@ -0,0 +1,39 @@
+; RUN: llc < %s -mtriple=s390x-linux-gnu | FileCheck %s
```
Comment (MaskRay) [inline llvm/test/CodeGen/SystemZ/xray-01.ll:1]: Perhaps just `xray.ll`

When we need other tests, they are likely suffixed, e.g. `xray-custom-log.ll`

**Review by MaskRay — APPROVED** [review]

**Comment by redstar** [comment]
@uweigand Just to make sure:

```
  bool HasVectorFeature =
      TM.getMCSubtargetInfo()->hasFeature(SystemZ::FeatureVector);
```

is the correct way to determine if the user compiles with the vector extension enabled, isn't it?

**Comment by uweigand** [comment]
> @uweigand Just to make sure:
> 
> ```
>   bool HasVectorFeature =
>       TM.getMCSubtargetInfo()->hasFeature(SystemZ::FeatureVector);
> ```
> 
> is the correct way to determine if the user compiles with the vector extension enabled, isn't it?

Yes, this should be correct.  I'm now wondering whether this handles -msoft-float correctly - for the main Subtarget, we disable the vector feature if soft-float is used, but I'm not sure if that also happens for the MCSubtarget.  Might be interesting to verify.

**Comment by redstar** [comment]
> Yes, this should be correct. I'm now wondering whether this handles -msoft-float correctly - for the main Subtarget, we disable the vector feature if soft-float is used, but I'm not sure if that also happens for the MCSubtarget. Might be interesting to verify.

That was a good point. It turns out that I had to check the soft-float feature in addition to the vector feature to get the correct result. I updated the test accordingly. Can you have another look, please?

**Review by uweigand — APPROVED** [review]
LGTM

**Pull request merged**